### PR TITLE
RESPMOD gets stuck with delay pools

### PR DESCRIPTION
**Pull request by eduard-bagdasaryan**
The transaction gets stuck if a delay pool individual speed is less than
~ 64K. HttpStateData::readReply() starts to wait until ModXact drains
its BodyPipe buffer but this never happens because
Icap::ModXact::virginConsume() returns early with the
"postponing consumption from" message. This does not happen if
the delay pool speed is high enough to fill the entire BodyPipe buffer,
because Icap::ModXact::virginConsume() proceeds to drain the buffer after
checking its "bp.buf().spaceSize() > 2" condition.

HttpStateData::readReply() should postpone reading only if the
BodyPipe buffer is full, instead of checking whether is not empty.

### CODE REVIEW

**Comment by eduard-bagdasaryan** [comment]
> rephrase to clarify the triggering condition 

Rephrased.

> Unchecked assertion: The current 2-byte condition in ModXact is more conservative than new/PR condition 

I agree with this assertion: to avoid stalling, http.cc code should block when the buffer is full and ModXact.cc should block (i.e., not consume) when the buffer is not full (that is obviously guaranteed by the buf().spaceSize() > 2 condition).

**Comment by eduard-bagdasaryan** [comment]
I also checked that Icap::ModXact::virginConsume() should not block on isRetriable flag for big responses (i.e., > 64K). Without previews, Icap::ModXact::decideOnRetries() disables retries because canBackupEverything() should be false in this case.

Also I tried to find out whether the 'bp.buf().spaceSize() > 2' condition in virginConsume() is still justified. Back in 478cfe9 (when it was added) it referenced HttpStateData::maybeReadVirginBody() method:
```
    /*
     * why <2? Because delayAwareRead() won't actually read if
     * you ask it to read 1 byte.  The delayed read request
     * just gets re-queued until the client side drains, then
     * the I/O thread hangs.  Better to not register any read
     * handler until we get a notification from someone that
     * its okay to read again.
     */
    if (read_sz < 2)
        return;
```
In that version, If one byte was passed to StoreEntry::delayAwareRead() and StoreEntry::bytesWanted(), it was treated as 0 bytes (due to suspicious aRange.end - 1 expressions there). However, I could not find these 'aRange.end - 1' expressions (or similar) in the modern delay_pools related code - instead, it operates with min/max ranges without any of 1 or 2 magic numbers. So It looks like the delay_pools code have been fixed since 478cfe9, and the '>2' condition in ModXact.cc seems stale. Note that http.cc code still has one '<2' check in HttpStateData::maybeMakeSpaceAvailable(), but it looks as a kind of optimization, avoiding unproductive 1-byte reads.
BTW, 19e97cf commit removed this '>2' condition on client side.

File: src/http.cc [hunk 1]
```diff
@@ -1216,7 +1216,7 @@ HttpStateData::readReply(const CommIoCbParams &io)
             return; // wait for Client::noteMoreBodySpaceAvailable()
         }
 
-        if (virginBodyDestination && virginBodyDestination->buf().hasContent()) {
+        if (virginBodyDestination && !virginBodyDestination->buf().hasPotentialSpace()) {
```
Comment (rousskov) [inline src/http.cc:1219]: > Alex: We will come back here after that PR merges.

PR 2056 got merged. Please reevaluate this condition and the corresponding condition in ModXact::virginConsume() condition. As we have [discussed](https://github.com/measurement-factory/squid/pull/291#pullrequestreview-2722073548), the two conditions should probably be the same (or, to be more precise, should match one another). FTP Client code may also have to be adjusted. Please adjust this PR description accordingly as well.

**Review by rousskov — CHANGES_REQUESTED** [review]

File: src/http.cc (hunk 1 above)
Comment (eduard-bagdasaryan) [reply src/http.cc:1219]: >  FTP Client code may also have to be adjusted

FTP Client does not need adjusting because it already has this condition in Client::calcBufferSpaceToReserve(). So if  adaptor_space is zero, calcBufferSpaceToReserve() returns 0 and Ftp::Client::maybeReadVirginBody() exits on the !read_sz condition (i.e., postponing until Client::noteMoreBodySpaceAvailable()). I also ran several tests with FTP requests (configuring delay pools and ICAP as before), and they worked without getting stuck.

**Comment by eduard-bagdasaryan** [comment]
Adjusted the PR description with the latest PR changes and removed XXXs.

**Review by rousskov — CHANGES_REQUESTED** [review]

File: src/clients/FtpClient.cc [hunk 2]
```diff
@@ -925,7 +925,7 @@ Ftp::Client::maybeReadVirginBody()
 
     initReadBuf();
 
-    const int read_sz = replyBodySpace(*data.readBuf, 0);
+    const auto read_sz = calcBufferSpaceToReserve(data.readBuf->spaceSize(), 0);
```
[the remark this answers is not in the document]
Comment (eduard-bagdasaryan) [reply src/clients/FtpClient.cc:928]: Yes, this improvement is not required for this PR. I created PR300 for it.

File: src/http.cc [hunk 3]
```diff
@@ -1216,7 +1216,7 @@ HttpStateData::readReply(const CommIoCbParams &io)
...
             debugs(11, 5, "avoid delayRead() to give adaptation a chance to drain body pipe buffer: " << virginBodyDestination->buf().contentSize());
             return; // wait for Client::noteMoreBodySpaceAvailable()
```
[the remark this answers is not in the document]
Comment (eduard-bagdasaryan) [reply src/http.cc:1221]: > We can reach this "give adaptation a chance" code without delay pools, right?

Theoretically - maybe (for some edge cases), but I could not do this in a test. For example, for responses >64K and a small read_ahead_gap (1K) - the entire buffer is filled at once (so ModXact does not block it). For bodies <64K it does not block either. Delay pools allow us not to read the first 64K all at once.

[the remark this answers is not in the document]
Comment (eduard-bagdasaryan) [reply src/http.cc:1221]: After more experiments and consideration I still think that this is not possible without delay pools.

In the test I tried to make `virginConsume()` not to consume the reply by pausing the ICAP server process from replying (because Squid calls `disableRetries()` when it receives the reply from ICAP). Replies >= 64K are out of consideration because Squid calls `disableRetries()` immediately if `ModXact::canBackupEverything()` returns false (`TheBackupLimit` precondition). So I experimented with replies < 64K, sending the body parts with small delays.

Avoiding `disableRetries()` proved to be trickier because Squid calls it in many other places, for example from `Xaction::openConnection()` if the connection is new (i.e., not pconn case). My test scenario worked when I started a transaction just after Squid initiated the OPTIONS ICAP request (and kept that pconn after that). However this transaction did not block because Squid's client part kept on reading the body. `HttpStateData::readReply()`
does not block because `readSizeWanted` never becomes 0.
This variable becomes 0 when `StoreEntry::bytesWanted()` returns 0, i.e., `mem_obj->readAheadPolicyCanRead()` is false, that is calculated as
```
 const bool canRead = endOffset() - savedHttpHeaders <  lowestMemReaderOffset() + Config.readAheadGap;
```
The point is that until ICAP replied, all the variables in this expression (except readAheadGap) remain 0 and so it always evaluates as true.
`readSizeWanted` may also become 0 ls if `HttpStateData::maybeMakeSpaceAvailable()` returns 0, but that means that the HTTP buffer is full. This, however, is not possible as long as the body pipe has enough space (i.e., we received less than 64K): the HTTP buffer is emptied once its bytes are copied to the body pipe.

[the remark this answers is not in the document]
Comment (rousskov) [reply src/http.cc:1221]: I understand that to make StoreEntry::bytesWanted() zero without delay pools, we need to fill read_ahead_gap. We cannot even start filling that gap until we have some adapted bytes (because _virgin_ response bytes are not added to StoreEntry when virginBodyDestination is set). Thus, Squid must receive (the beginning of) an ICAP response body and place it in StoreEntry for the test to work.

As you have pointed out, having adapted bytes implies disableRetries(). OK.

What I do not understand yet is why we need to avoid disableRetries(). ModXact::virginConsume() "postponing consumption" code works when retries are disabled (i.e. isRetriable is false):

```
    if (isRetriable)
        return; // do not consume if we may have to retry later

    BodyPipe &bp = *virgin.body_pipe;
    const bool wantToPostpone = isRepeatable || canStartBypass || protectGroupBypass;

    if (wantToPostpone && bp.buf().spaceSize() > 0) {
        // Postponing may increase memory footprint and slow the HTTP side
        // down. Not postponing may increase the number of ICAP errors
        // if the ICAP service fails. We may also use "potential" space to
        // postpone more aggressively. Should the trade-off be configurable?
        debugs(93, 8, "postponing consumption from " << bp.status());
        return;
    }
```

Thus, we want true wantToPostpone (when isRetriable is already false). 

`Icap::ModXact::virginConsume() returns early with the "postponing consumption from" message` part of PR description implies that the above debugs() is reachable (at least with delay pools). In other words, wantToPostpone was true (in some tests with delay pools). On the surface, wantToPostpone condition is not directly influenced by delay pools; it is more about ICAP configuration and state. Can wantToPostpone be true (in certain ICAP configurations) without delay pools?

[the remark this answers is not in the document]
Comment (eduard-bagdasaryan) [reply src/http.cc:1221]: > why we need to avoid disableRetries()

I forgot to mention that I needed this in my test to make Squid reuse the existing connection to the forked icap_server process (which I could stop by a signal beforehand). It is clear that `virginConsume()` may postpone consumption due to other flags as well.

> Can wantToPostpone be true (in certain ICAP configurations) without delay pools?

Yes, of course, but this postponing stops when Squid receives headers from ICAP server, calling `ModXact::startSending()` and disable*() methods that clear those flags used for wantToPostpone calculation.

[the remark this answers is not in the document]
Comment (rousskov) [reply src/http.cc:1221]: > > Can wantToPostpone be true (in certain ICAP configurations) without delay pools?
 
> Yes, of course, but this postponing stops when Squid receives headers from ICAP server, calling `ModXact::startSending()` and disable*() methods that clear those flags used for wantToPostpone calculation.

Got it, thank you! I have adjusted PR description to include the corresponding explanation.

I also removed references to specific delay pools speed settings. AFAICT, the stall is possible with nearly any delay pool speed limit as long as Squid manages to read more data than that setting allows (but still less than the buffer capacity).

If the current PR description looks OK to you, then we are done here!

**Review by rousskov — APPROVED** [review]
